App/Services/UriHandler.cs

```diff
@@ -0,0 +1,150 @@
+using System;
+using System.Collections.Specialized;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Web;
+using Coder.Desktop.App.Models;
+using Coder.Desktop.Vpn.Proto;
+using Microsoft.Extensions.Logging;
+
+
+namespace Coder.Desktop.App.Services;
+
+public interface IUriHandler : IAsyncDisposable
+{
+    public Task HandleUri(Uri uri, CancellationToken ct = default);
+}
+
+public class UriHandler(
+    ILogger<UriHandler> logger,
+    IRpcController rpcController,
+    IUserNotifier userNotifier,
+    IRdpConnector rdpConnector) : IUriHandler
+{
+    private const string OpenWorkspacePrefix = "/v0/open/ws/";
+
+    internal class UriException(string title, string detail) : Exception
+    {
+        internal readonly string Title = title;
+        internal readonly string Detail = detail;
+    }
+
+    public async Task HandleUri(Uri uri, CancellationToken ct = default)
+    {
+        try
+        {
+            await HandleUriThrowingErrors(uri, ct);
+        }
+        catch (UriException e)
+        {
+            await userNotifier.ShowErrorNotification(e.Title, e.Detail, ct);
+        }
+    }
+
+    private async Task HandleUriThrowingErrors(Uri uri, CancellationToken ct = default)
+    {
+        if (uri.AbsolutePath.StartsWith(OpenWorkspacePrefix))
+        {
+            await HandleOpenWorkspaceApp(uri, ct);
+            return;
+        }
+
+        logger.LogWarning("unhandled URI path {path}", uri.AbsolutePath);
+        throw new UriException("URI handling error",
+            $"URI with path {uri.AbsolutePath} is unsupported or malformed");
```

Comment:
I think in a lot of these error messages it might be nice to surround the templated variables with single quotes. It can help users determine when a string that was supposed to be specified was empty, otherwise it's as if it's not in the string at all.